### ISSUE TITLE
Space removed after annotation - Want to keep it

### ISSUE THREAD

**Issue by d1egoaz** [scalameta/scalafmt#618]
When using [Play DI Guice](https://www.playframework.com/documentation/2.5.x/ScalaDependencyInjection#Declaring-runtime-DI-dependencies), I would like to keep the space after the `@Inject` annotation

**Version**: `scalafmt 0.5.0-RC2`
**Integration**: cli
**Configuration**: `style=defaultWithAlign`

```scala
class Foo @Inject() (conf: Configuration)
```

Scalafmt:
```scala
class Foo @Inject()(conf: Configuration)
```

I would like the input to look like this:
```scala
class Foo @Inject() (conf: Configuration)
```

I think I can help with a PR if you could point me in the right direction/where to start.

And thank you for this amazing project 🍻 , I've been using it for a few weeks now, it has been really useful since I moved from IntelliJ to Emacs

**Comment by olafurpg** [comment]
Thanks for reporting! If you are motivated to take a stab at this, you can track down where the `NoSplit` (no space) origins between `)(` using the tutorial in the readme: https://github.com/olafurpg/scalafmt#tutorial-add-spacesinparentheses-flag

Each whitespace token (`NoSplit`, `Space` or `Newline`) between non-whitespace tokens can be tracked down to a line number in Router.scala.

The tutorial is slightly outdated, please let me know if there's anything confusing!

**Comment by AnamikaD** [comment]
@olafurpg I ran "sbt test"
I get this error (screenshot attached). Tried googling with no success.
![screenshot from 2017-03-02 22-46-17](https://cloud.githubusercontent.com/assets/15039155/23518674/42b1343c-ff9a-11e6-93f6-14d2e35e5525.png)

**Comment by olafurpg** [comment]
@AnamikaD Great you got the tests running! That intellij/compile failure is not your fault, it's a bug in the build that I fix in https://github.com/olafurpg/scalafmt/pull/760 To get `intellij/compile` to work, you need to run `intellij/updateIdea` first, and that will download a **lot** of files (~800mb)  that are only relevant when developing the intellij plugin.

**Comment by AnamikaD** [comment]
@olafurpg  Thank you for the reply! I will go forward with the code and other tasks given.

**Comment by AnamikaD** [comment]
@olafurpg According to this: https://github.com/olafurpg/scalafmt/blob/master/Tutorial.md, In Router.scala line 1276 contains " Split(NoSplit, 0) ", but I see this: " Split(Space, 0, ignoreIf = newlines > 0) "!

**Comment by olafurpg** [comment]
@AnamikaD Yes, that can very likely be. The tutorial was last updated in early January and Router.scala has changed a bit since then. If you run the tests yourself, you should be able to see the new line numbers.

**Comment by AnamikaD** [comment]
@olafurpg  Thanks! Got it.

**Comment by AnamikaD** [comment]
@olafurpg 
In Hacking.stat:  
maxColumn = 100
<<< ONLY Spaces in after close braces
class Foo @Inject()(conf: Configuration)
>>>
class Foo @Inject() (conf: Configuration)

In Router.scala, line 356 added this,
if (style.spaces.afterCloseBraces) Space
              else NoSplit 
afterCloseBraces is defined true in Space.scala. Test case failed. Any leads/Guide??

**Comment by olafurpg** [comment]
@AnamikaD Please try to wrap code block with triple backticks `, that makes it easier to read.

Have you tried to set 

```
spaces.afterCloseBraces = true
```

after the `maxColumns` line? 

BTW. I don't think there's a need to introduce a new flag for this change, we can include the change in the default style.

**Comment by AnamikaD** [comment]
@olafurpg  sorry for the trouble.
Hacking.stat code:
``` maxColumn = 100
spaces.afterCloseBraces = true
<<< ONLY Spaces in after close braces
class Foo @Inject()(conf: Configuration)
>>>
class Foo @Inject() (conf: Configuration) ```

Failed test case. Where is this default style? do we need to check whether there is an '(' after ')'?

**Comment by olafurpg** [comment]
The default style is the configuration that is used everywhere unless you override settings with for example `maxColumn = 100`.

That test looks good, except you can remove the `spaces.afterCloseBraces = true` part. I think we want to handle the case [here](https://github.com/olafurpg/scalafmt/blob/5366b383a5568fbba8633c9096932919c100e685/core/src/main/scala/org/scalafmt/internal/Router.scala#L350-L355) so that the split is a `Space` when `leftOwner.parent.exists(_.is[Mod])`

**Comment by AnamikaD** [comment]
@olafurpg  what is leftOwner and Mod? what does it mean?

**Comment by AnamikaD** [comment]
I get the output as 
` 
class Foo @Inject() (
  conf: Configuration )
`
The new line after '(' is caused by,
[here](https://github.com/olafurpg/scalafmt/blob/5366b383a5568fbba8633c9096932919c100e685/core/src/main/scala/org/scalafmt/internal/Router.scala#L702)

**Comment by olafurpg** [comment]
@AnamikaD 

> what is leftOwner and Mod? what does it mean?

`leftOwner` is the "owner" or "parent" of the left token. Scalafmt views code as an abstract syntax tree where the leafs of the tree are tokens, something like this 

![](https://i-msdn.sec.s-msft.com/dynimg/IC776861.png)

When formatting something like `class Foo @Inject() (x: Int)`, scalafmt has the task of deciding what to but between non-whitespace token, something like filling the `_` parts with space/newline/empty characters


```
     _   _ _      _ _ _
class Foo @ Inject ( ) (
```

`Mod` is a modifier tree node in the abstract syntax tree, for example like `@Inject()`. To learn more about syntax trees and tokens in scala, you might want to go through http://scalameta.org/tutorial/ I can highly recommend installing the [Ammonite repl](http://www.lihaoyi.com/Ammonite/#Ammonite-REPL) and play around with how the syntax tree of different code looks like,

```scala
$ amm // start ammonite repl from terminal console
Loading...
Welcome to the Ammonite Repl 0.8.2
(Scala 2.12.1 Java 1.8.0_102)
@ import $ivy.`org.scalameta:contrib_2.12:1.6.0`, scala.meta._, contrib._ // this loads the scala.meta dependency
import $ivy.$                                 , scala.meta._, contrib._
@ q"class Foo @Inject() (x: Int)".structure // this command prints out the abstract syntax tree of that code
res1: String = """
Defn.Class(Nil, Type.Name("Foo"), Nil, Ctor.Primary(Seq(Mod.Annot(Term.Apply(Ctor.Ref.Name("Inject"), Nil))), Ctor.Ref.Name("this"), Seq(Seq(Term.Param(Nil, Term.Name("x"), Some(Type.Name("Int")), None)))), Template(Nil, Nil, Term.Param(Nil, Name.Anonymous(), None, None), None))
"""  // Notice the `Mod.Annot` part
```

**Comment by AnamikaD** [comment]
@olafurpg Good explanation! I will walk through the tutorials asap. Can you tell me why are we getting a new line after '@Inject() (' ? (my previous comment).

**Comment by olafurpg** [comment]
@AnamikaD The tutorial explain how to trace a newline/space down to a line number in Router.scala, maybe the `maxColumn` is set too low?

**Comment by AnamikaD** [comment]
@olafurpg previously, went through the line number, made few changes still the output was incorrect. So after increasing maxColumn, it worked though I have 5 test cases failed. Attached screen shot
![screenshot from 2017-03-08 19-18-01](https://cloud.githubusercontent.com/assets/15039155/23706514/80cfc19a-0434-11e7-93bb-cf253aee8ddc.png)
.

**Comment by AnamikaD** [comment]
#835

**Comment by AnamikaD** [comment]
#Fix #618, add space after annotation #835

**Comment by olafurpg** [comment]
Closing due to inactivity, please reopen if you would like to take a stab at this!
